plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java

```diff
@@ -62,71 +62,107 @@ public TableChangesSplitSource(
+        currentVersion = new AtomicLong(extractCurrentVersion(functionHandle, logDir, fileSystem));
         tableReadVersion = functionHandle.tableReadVersion();
-        transactionLogDir = getTransactionLogDir(tableLocation);
+        transactionLogDir = logDir;
+    }
+
+    private static long extractCurrentVersion(TableChangesTableFunctionHandle functionHandle, String transactionLogDir, TrinoFileSystem fileSystem)
+    {
+        if (functionHandle.firstReadVersion().isPresent()) {
+            return functionHandle.firstReadVersion().get();
+        }
+        long currentVersion = functionHandle.tableReadVersion();
+        while (currentVersion >= 0) {
+            Location transactionLogFilePath = getTransactionLogJsonEntryPath(transactionLogDir, currentVersion);
+            try {
+                if (fileSystem.newInputFile(transactionLogFilePath).exists()) {
+                    List<DeltaLakeTransactionLogEntry> entries = getEntries(currentVersion, transactionLogDir, fileSystem);
+                    if (entries.isEmpty()) {
+                        throw new TrinoException(DELTA_LAKE_BAD_DATA, "Delta log should not contain empty entry");
+                    }
+                    CommitInfoEntry commitInfo = extractCommitInfoEntry(entries);
+                    if (commitInfo.getTimestamp() < functionHandle.firstReadTimestamp().get()) {
+                        return currentVersion + 1;
+                    }
+                }
+            }
+            catch (IOException exception) {
+                throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", exception);
+            }
+            currentVersion--;
+        }
+        return 0; // provided sinceTimestamp is so much in the past that we need to read entire table
+    }
+
+    private static List<DeltaLakeTransactionLogEntry> getEntries(long currentVersion, String transactionLogDir, TrinoFileSystem fileSystem)
+    {
+        try {
+            return getEntriesFromJson(currentVersion, transactionLogDir, fileSystem)
+                    .orElseThrow(() -> new TrinoException(DELTA_LAKE_BAD_DATA, "Delta Lake log entries are missing for version " + currentVersion));
+        }
+        catch (IOException e) {
+            throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+        }
+    }
+
+    private static CommitInfoEntry extractCommitInfoEntry(List<DeltaLakeTransactionLogEntry> entries)
+    {
+        return entries.stream()
+                .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseThrow();
     }
 
     @Override
     public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize) // TODO dont ignore maxSize https://github.com/trinodb/trino/issues/17182
     {
         long processedVersion = currentVersion.getAndIncrement();
-        try {
```

Comment:
Why try catch is removed?

Reply:
It is not removed, that code is moved a method that is called inside constructor.  
Now with timestamp the currentVersion is set based on firstReadVersion(if present) or is calculated based on firstReadTimestamp so I thought it was better to calculate it in constructor